### PR TITLE
feat(react-grid): rework virtual table

### LINKED ISSUE #33: Column width is changing on scrolling a virtual table the first time
Steps to reproduce:

1. Initialize DataGrid with empty rows;
2. Set new rows, at least 100. It should force the virtual table to render the vertical scrollbar;
3. Scroll the virtual table.

### LINKED ISSUE #280: Virtual Table Height
Is there a way to set the height for the container div of virtual table component via its props, instead of its fixed 530px height ?

### LINKED ISSUE #401: Feature Request: VirtualTableView for Material-UI
Hi, is it posslible to do scrollable table body in material-ui theme?

### PR DESCRIPTION
**Pull request by kvet**
fixes #33, closes #280, closes #401

### CODE REVIEW

File: packages/dx-react-demos/src/material-ui/featured-virtual-scrolling/demo.jsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        />
+
+        <LocalFiltering />
+        <LocalSorting />
+        <LocalGrouping />
+
+        <SelectionState />
+
+        <VirtualTableView
+          tableCellTemplate={({ row, column, style }) => {
```
Comment (SergeyAlexeev) [inline packages/dx-react-demos/src/material-ui/featured-virtual-scrolling/demo.jsx]: Extract template to a separate function.

File: packages/dx-react-grid/src/components/table-layout/virtual-table-utils.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
+const OVERSCAN = 3;
+
+const STUB_TYPE = 'stub';
+const VISIBLE_TYPE = 'visible';
+const OVERSCAN_TYPE = 'overscan';
+
+const STARTING_KEY = 'starting';
+const ENDING_KEY = 'ending';
+
+export const getVisibleRows = (rows, top, height, getRowHeight) => {
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/src/components/table-layout/virtual-table-utils.js]: The `height` arg is used only for the `bottom` calculation. Can we pass `bottom` to `getVisibleRows`? Then, we can replace `const size = getRowHeight(row);` to the `const height = getRowHeight(row);`. It's more eloquent.

File: packages/dx-react-grid/src/components/table-layout/virtual-table-utils.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+export const firstVisibleRowOffset = (prevVisibleRows, visibleRows) => {
+  const firstVisibleRowIndex = visibleRows.findIndex(row => row.type === VISIBLE_TYPE);
+  if (firstVisibleRowIndex === -1) return 0;
+
+  const firstVisibleRow = visibleRows[firstVisibleRowIndex].row;
+  const prevIndex = prevVisibleRows.findIndex(row => row.row === firstVisibleRow);
+  if (prevIndex === -1) return 0;
+
+  const position = visibleRows
```
Comment (SergeyAlexeev) [inline packages/dx-react-grid/src/components/table-layout/virtual-table-utils.js]: Seems like, `const position = visibleRows...` and `const prevPosition = prevVisibleRows...` calculations are equal and can be extracted to the `calculatePosition` function, for instance.

File: packages/dx-react-demos/src/bootstrap3/templates/progress-bar-cell.jsx [hunk 4]
```diff
@@ -7,7 +7,7 @@ import {
 } from 'react-bootstrap';
 
 export const ProgressBarCell = ({ value, style }) => (
-  <td style={style}>
+  <td style={{ position: 'relative', ...style }}>
```
[the remark this answers is not in the document]
Comment (kvet) [reply packages/dx-react-demos/src/bootstrap3/templates/progress-bar-cell.jsx]: There is a rendering problem in Safari without this style.

File: packages/dx-react-grid/docs/reference/virtual-table-view.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+### Dependencies
+
+The plugin requires dependencies from the [TableView](table-view.md#dependencies) plugin.
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+height | number | 530 | A number used to specify height of virtual table.
+estimatedRowHeight | number | `37` for [Bootstrap3](https://www.npmjs.com/package/@devexpress/dx-react-grid-bootstrap3) and `48` for [Material UI](https://www.npmjs.com/package/@devexpress/dx-react-grid-material-ui) | A number used as a row height before a row is rendered. Providing a precise estimate of the height of rows can improve the performance of the virtual table. If the table contains variable height rows, you should specify a median row height value.
```
[the remark this answers is not in the document]
Comment (kvet) [reply packages/dx-react-grid/docs/reference/virtual-table-view.md]: You are right. It a valuable property. Like the `height` property. But I don't think that we should add a demo for a property which work can't be visible. The documentation says that this 'can improve performance'. It is not guaranteed.

**Pull request merged**